### PR TITLE
SIL serialization and IRGen emission improvements

### PR DESCRIPTION
**Pull request by swiftix**
- Try to avoid increasing the visibility of SILFunctions in IRGen when emitting LLVM IR for them

   Any non-external SIL functions should preserve their linkage, instead of making them `public` in the object file as we did before. This means that clients cannot reference them anymore, which is a good thing. Client should only reference public symbols from the compiled module.

   When it comes to external non-public functions, change their linkage to shared_external and later emit them into the client object file, instead of assuming these functions are available elsewhere.

- Do not SIL serialize bodies of `@_semantics("stdlib_binary_only")` functions

- No need to run `ExternalDefsToDecls` at -Onone and `ExternalFunctionDefinitionsElimination` at -O anymore. The passes are not removed in this commit yet. It will be done later.

### CODE REVIEW

**Comment by swiftix** [comment]
@swift-ci please smoke test

**Comment by swiftix** [comment]
@swift-ci please smoke test

**Comment by swiftix** [comment]
@swift-ci please smoke benchmark

File: lib/IRGen/GenDecl.cpp [hunk 1]
```diff
@@ -1353,7 +1353,7 @@ bool LinkEntity::isFragile(ForDefinition_t isDefinition) const {
 static std::tuple<llvm::GlobalValue::LinkageTypes,
                   llvm::GlobalValue::VisibilityTypes,
                   llvm::GlobalValue::DLLStorageClassTypes>
-getIRLinkage(const UniversalLinkageInfo &info, SILLinkage linkage,
+getIRLinkage(const UniversalLinkageInfo &info, const LinkEntity *entity, SILLinkage linkage,
```
Comment (slavapestov) [inline lib/IRGen/GenDecl.cpp]: I was imagining isFragile would go away altogether?

File: lib/IRGen/IRGenSIL.cpp [hunk 2]
```diff
@@ -1460,6 +1460,11 @@ static void emitLocalSelfMetadata(IRGenSILFunction &IGF) {
 void IRGenModule::emitSILFunction(SILFunction *f) {
   if (f->isExternalDeclaration())
     return;
+  // Do not emit bodies of public_external functions.
+  // The only exception is transparent functions.
+  if (hasPublicVisibility(f->getLinkage()) && f->isAvailableExternally() &&
+      !f->isTransparent())
```
Comment (slavapestov) [inline lib/IRGen/IRGenSIL.cpp]: Transparent functions should have shared linkage instead, then you wouldn't need this special case

File: lib/SIL/SILDeclRef.cpp [hunk 3]
```diff
@@ -381,6 +381,15 @@ SILLinkage SILDeclRef::getLinkage(ForDefinition_t forDefinition) const {
   // FIXME: @objc declarations should be too, but we currently have no way
   // of marking them "used" other than making them external. 
   ValueDecl *d = getDecl();
+
+  // stdlib_binary_only functions should have a public SIL linkage.
+  auto Attrs = d->getAttrs();
+  for (auto *A : Attrs.getAttributes<SemanticsAttr>()) {
+    if (A->Value == "stdlib_binary_only") {
+      return maybeAddExternal(SILLinkage::Public);
```
Comment (slavapestov) [inline lib/SIL/SILDeclRef.cpp]: Are they not declared as public in the source? Why not?

File: lib/SILGen/SILGen.cpp [hunk 4]
```diff
@@ -972,13 +974,13 @@ SILFunction *SILGenModule::emitLazyGlobalInitializer(StringRef funcName,
...
-                       isMakeModuleFragile()
-                           ? IsSerialized
-                           : IsNotSerialized);
+  auto *f = M.createFunction(
+      isMakeModuleFragile() ? SILLinkage::Public : SILLinkage::Private,
+      funcName, initSILType, nullptr, SILLocation(binding), IsNotBare,
+      IsNotTransparent, isMakeModuleFragile() ? IsSerialized : IsNotSerialized);
+  // Do not serialize the initializer in the -sil-serialize-all mode.
+  if (isMakeModuleFragile())
+    f->addSemanticsAttr("stdlib_binary_only");
```
Comment (slavapestov) [inline lib/SILGen/SILGen.cpp]: Why not just pass IsNotSerialized to the constructor above?

File: lib/SILOptimizer/Transforms/FunctionSignatureOpts.cpp [hunk 5]
```diff
@@ -681,13 +681,21 @@ void FunctionSignatureTransform::createFunctionSignatureOptimizedFunction() {
...
+  // Create the thunk body!
   F->setThunk(IsThunk);
   // The thunk now carries the information on how the signature is
   // optimized. If we inline the thunk, we will get the benefit of calling
   // the signature optimized function without additional setup on the
   // caller side.
   F->setInlineStrategy(AlwaysInline);
+  // If the new callee is @_semantics("stdlib_binary_only"), it is fine to call
+  // it, because it will be exposed as a public symbol in the object file.
+  if (F->hasSemanticsAttr("stdlib_binary_only")) {
```
Comment (slavapestov) [inline lib/SILOptimizer/Transforms/FunctionSignatureOpts.cpp]: This should instead be encoded as the function having IsNotSerialized set. Why do we need to specially handle it here?

File: lib/SIL/SILDeclRef.cpp (hunk 3 above)
Comment (swiftix) [reply lib/SIL/SILDeclRef.cpp]: We have a number of `@_semantics("stdlib_binary_only")` uses in the stdlib, which are not public. It seems like this attribute was considered to be totally orthogonal to the access level. It is kind of similar to the effect of using `@_versioned` in the sense that the object file would contain it as a public symbol.

**Comment by CodaFi** [comment]
This appears to be superseded by 6f3b326d121a24f1fa4473a049d7b53d1ee2eaa7

**Pull request closed without merging**